### PR TITLE
fix #11354 jsgen not carefully handle genAddr with nkHiddenAddr,nkStm…

### LINKED ISSUE #11354: JS compiler error related to implicit return and return var type
### Example

```nim
type
  TrackySeq[T] = object
    s: seq[T]
    pos: int

proc foobar(ls: var TrackySeq[seq[int]], i: int): var seq[int] =
  echo ls.pos  # removing this, or making the return explicit works
  ls.s[i]

var foo: TrackySeq[seq[int]]
foo.s.add(@[0])
foo.foobar(0).add(1)
echo foo.s
```

### Current Output
```
foob.nim(8, 3) Error: internal error: genAddr: nkStmtListExpr
```

### Expected Output
```
0
@[@[0, 1]]
```

```
$ nim -v
Nim Compiler Version 0.19.9 [Linux: amd64]
Compiled at 2019-05-18
Copyright (c) 2006-2019 by Andreas Rumpf

active boot switches: -d:release
```

### LINKED ISSUE #14574: js: `lent` doesn't work in iterators with tuple destructuring
`toTable` doesn't work with js target with values of non-primitive type: Cannot read property 'slice' of undefined
### Example
```nim
import tables
type Foo = object
  a: int
proc fn() =
  let x = Foo(a: 10) # likewise with tuples
  block:
    var t: Table[string, Foo]
    t["abc"] = x # ok
  block:
    let t = {"abc": x}.toTable # bug
fn()
```

### Current Output
nim js -r --lib:lib $timn_D/tests/nim/all/t10893.nim
```
/Users/timothee/git_clone/nim/timn/tests/nim/all/t10893.js:170
        result_33250219 = src_33245024.slice(0);
                                       ^

TypeError: Cannot read property 'slice' of undefined
    at nimCopy (/Users/timothee/git_clone/nim/timn/tests/nim/all/t10893.js:170:40)
    at nimCopyAux (/Users/timothee/git_clone/nim/timn/tests/nim/all/t10893.js:1994:48)
    at nimCopyAux (/Users/timothee/git_clone/nim/timn/tests/nim/all/t10893.js:1999:7)
    at nimCopy (/Users/timothee/git_clone/nim/timn/tests/nim/all/t10893.js:145:7)
    at toTable_34360655 (/Users/timothee/git_clone/nim/timn/tests/nim/all/t10893.js:2647:13)
    at fn_34271017 (/Users/timothee/git_clone/nim/timn/tests/nim/all/t10893.js:2673:24)
    at Object.<anonymous> (/Users/timothee/git_clone/nim/timn/tests/nim/all/t10893.js:2682:1)
    at Module._compile (internal/modules/cjs/loader.js:1200:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1220:10)
    at Module.load (internal/modules/cjs/loader.js:1049:32)
```

### Expected Output
```
works
```

### Additional Information
* devel 0824fdd06100282a853fa50efb66e2161f4235d8 1.3.5

EDIT:
this works:
```nim
proc toTable2*[A, B](pairs: openArray[(A, B)]): Table[A, B] =
  result = initTable[A, B](rightSize(pairs.len))
  # for key, val in items(pairs):
  for i in 0..<len(pairs):
    result[pairs[i][0]] = pairs[i][1]
```

so looks like what's broken is items tuple destructuring:

```nim
  proc fn() =
    let a = @[("foo", (12, 13))]
    for (k,v) in a:
      echo (k,v)
  fn()
```
crashes with:
```
/Users/timothee/git_clone/nim/timn/tests/nim/all/t10893.js:198
        result_33250219 = src_33245024.slice(0);
                                       ^

TypeError: src_33245024.slice is not a function
    at nimCopy (/Users/timothee/git_clone/nim/timn/tests/nim/all/t10893.js:198:40)
    at nimCopyAux (/Users/timothee/git_clone/nim/timn/tests/nim/all/t10893.js:366:48)
    at nimCopyAux (/Users/timothee/git_clone/nim/timn/tests/nim/all/t10893.js:371:7)
    at nimCopy (/Users/timothee/git_clone/nim/timn/tests/nim/all/t10893.js:173:7)
    at fn_33627401 (/Users/timothee/git_clone/nim/timn/tests/nim/all/t10893.js:895:13)
    at Object.<anonymous> (/Users/timothee/git_clone/nim/timn/tests/nim/all/t10893.js:915:1)
    at Module._compile (internal/modules/cjs/loader.js:1200:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1220:10)
    at Module.load (internal/modules/cjs/loader.js:1049:32)
    at Function.Module._load (internal/modules/cjs/loader.js:937:14)
```

EDIT: bug disappears if I remove `lent` in iterators.nim
```nim
iterator items*[T](a: seq[T]): lent2 T {.inline.} =
=>
iterator items*[T](a: seq[T]): T {.inline.} =
```

### PR DESCRIPTION
**Pull request by bung87**
…tListExpr; genAsgn with lvalue tyVar and rvalue tyPtr

this also partial fix https://github.com/nim-lang/Nim/issues/14574

### CODE REVIEW

File: compiler/jsgen.nim [hunk 1]
```diff
@@ -1319,6 +1323,10 @@ proc genAddr(p: PProc, n: PNode, r: var TCompRes) =
     gen(p, n[0], r)
   of nkHiddenDeref:
     gen(p, n[0][0], r)
+  of nkHiddenAddr:
+    gen(p, n[0], r)
+  of nkStmtListExpr:
+    gen(p, n[^1], r)
```
Comment (Araq) [inline compiler/jsgen.nim]: Incorrect for ``addr((echo "unrelated"; x))``. (But maybe the frontend doesn't allow this?)

Comment (bung87) [reply compiler/jsgen.nim]: oh,  so the ast results `results = addr((echo "unrelated"; x))`, how I correct handle this ? the cgen has logic turn statement list expression to statement list

Comment (Araq) [reply compiler/jsgen.nim]: For now ensure ``n.len == 1`` and if not, produce the previous internal error.

File: compiler/jsgen.nim [hunk 2]
```diff
@@ -1318,7 +1322,13 @@ proc genAddr(p: PProc, n: PNode, r: var TCompRes) =
   of nkObjDownConv:
     gen(p, n[0], r)
   of nkHiddenDeref:
-    gen(p, n[0][0], r)
+    gen(p, n[0], r)
+  of nkHiddenAddr:
+    if not n.len == 1:
```
Comment (Araq) [inline compiler/jsgen.nim]:
```suggestion 
    if n.len != 1
```

File: compiler/jsgen.nim [hunk 3]
```diff
@@ -1318,7 +1322,13 @@ proc genAddr(p: PProc, n: PNode, r: var TCompRes) =
   of nkObjDownConv:
     gen(p, n[0], r)
   of nkHiddenDeref:
-    gen(p, n[0][0], r)
+    gen(p, n[0], r)
+  of nkHiddenAddr:
+    if n.len != 1
```
Comment (Araq) [inline compiler/jsgen.nim]:
```suggestion
    if n.len != 1:
```

File: compiler/jsgen.nim [hunk 4]
```diff
@@ -1318,7 +1322,13 @@ proc genAddr(p: PProc, n: PNode, r: var TCompRes) =
   of nkObjDownConv:
     gen(p, n[0], r)
   of nkHiddenDeref:
-    gen(p, n[0][0], r)
+    gen(p, n[0], r)
+  of nkHiddenAddr:
+    if n.len != 1:
+      internalError(p.config, n[0].info, "genAddr: " & $n[0].kind)
```
Comment (Araq) [inline compiler/jsgen.nim]:
```suggestion
```

File: compiler/jsgen.nim [hunk 5]
```diff
@@ -1318,7 +1322,13 @@ proc genAddr(p: PProc, n: PNode, r: var TCompRes) =
...
+    gen(p, n[0], r)
+  of nkStmtListExpr:
+    gen(p, n[0], r)
```
Comment (Araq) [inline compiler/jsgen.nim]:
```suggestion
    if n.len == 1: gen(p, n[0], r)
    else: internalError(p.config, n[0].info, "genAddr for complex nkStmtListExpr")
```

**Comment by timotheecour** [comment]
@bung87 please be careful, when writing `partial fix #14574`, github interprets it as `fix #14574` and auto-closes it when your PR is merged; but the bug is still there in #14574

**Comment by bung87** [comment]
> @bung87 please be careful, when writing `partial fix #14574`, github interprets it as `fix #14574` and auto-closes it when your PR is merged; but the bug is still there in #14574

good to know , how is that issue goes on ? have you tried fix the remain problems ? maybe I can give a try

**Comment by timotheecour** [comment]
I haven't made progress on it besides https://github.com/nim-lang/Nim/pull/14577 which only fixed 1 issue mentioned there but not the rest

>  maybe I can give a try

great

**Pull request merged**